zerver/actions/message_send.py

```diff
@@ -858,11 +863,118 @@ def get_active_presence_idle_user_ids(
+        col_name: getattr(message, col_name) for col_name in field_names
+    }
+
+    # The goal of the following query is to implement Idempotency for message creation,
+    # and that only one request can do the work at a time.
+
+    # The final SELECT combines data from the 3 CTEs and
+    # returns 3 deterministic values in the following order:
+    # 1- status: Indicatesa a NEW or DUPLICATE message.
+    # 2- new_msg_id: The DB-generated id of the inserted Message (None in case of no insertion).
+    # 3- cached_response: The saved result returned as the API response for duplicate message.
+
+    # Example output:
+    # new message: ('NEW', 315, None)
+    # duplicate: ('DUPLICATE', None, '{"message_id": 313, "automatic_new_visibility_policy": 3}')
+
+    # CTEs:
+    # 1.status_cte:
+    # The status of message insertion, new or duplicate.
+    # MATERIALIZED to preserve the state across the rest of query.
+    # Locks the row, but abort if another request is already working.
+
+    # 2.insertion_cte:
+    # References status_cte to conditionally insert a message
+    # for a new message (status_cte.identifier IS NULL).
+    # Returns the new db-generated id of that inserted message (null in case of no insertion).
+
+    # 3.update_cte:
+    # References insertion_cte to conditionally SET insertion_cte.identifier to the new inserted message.
+    # Although there is no explicit condition (WHERE), nothing actually updates
+    # if insertion_cte returns null
+
+    with connection.cursor() as cursor:
+        query = SQL(
+            """
+            WITH status_cte AS MATERIALIZED (
+                SELECT * FROM {idempotency_table}
+                WHERE idempotency_key = %(idempotency_key_value)s
+                FOR UPDATE NOWAIT
+            ),
+            insertion_cte AS(
+                INSERT INTO {message_table} ({msg_cols})
+                SELECT {values_placeholders}
+                WHERE (SELECT identifier FROM status_cte) IS NULL
+                RETURNING {message_pk} AS new_msg_id
+            ),
+            update_cte AS(
+                UPDATE {idempotency_table}
+                SET identifier = insertion_cte.new_msg_id
+                FROM insertion_cte
+                WHERE idempotency_key = %(idempotency_key_value)s
+            )
+            SELECT
+                CASE
+                    WHEN (SELECT identifier FROM status_cte) IS NULL THEN 'NEW' ELSE 'DUPLICATE'
+                END AS status,
+                (SELECT new_msg_id FROM insertion_cte),
+                (SELECT cached_response FROM status_cte);
+        """
+        ).format(
+            message_table=Identifier(Message._meta.db_table),
+            idempotency_table=Identifier(IdempotentRequest._meta.db_table),
+            msg_cols=SQL(", ").join(map(Identifier, field_names)),
+            values_placeholders=SQL(", ").join(map(Placeholder, field_names)),
+            message_pk=Identifier(Message._meta.pk.name),
+        )
```

Comment:
Insofar as you think this code will be reusable outside the context of handling Message rows in particular, I think you can optimistically make some things generic.  For example, I would replace message_table and msg_cols and message_pk with just `table` and `cols` and `message_pk` as the format keys.  And I would introduce local vars:

~~~ py
table = Identifier(Message._meta.db_table)
pk = Identifier(Message._meta.pk.name)

// ...
    ..format(
        table=table,
        pk=pk,
        # ...
    )